lib/logstash/plugin_mixins/event_support/from_json_helper.rb

```diff
@@ -0,0 +1,49 @@
+# encoding: utf-8
+
+require 'logstash/plugin_mixins/event_support'
+require 'logstash/plugin_mixins/event_support/event_factory_adapter'
+
+require 'logstash-core'
+require 'logstash/event'
+
+module LogStash
+  module PluginMixins
+    module EventSupport
+      module FromJsonHelper
+
+        def self.included(base)
+          if defined? LogStash::Plugins::EventFactorySupport # native support from LS core
+            base.send(:include, NativeImpl)
+          else
+            require 'logstash/json'
+            base.send(:include, FallbackImpl)
+          end
+        end
+
+        module NativeImpl
+
+          def events_from_json(json, event_factory)
+            LogStash::Event.from_json(json) { |data| event_factory.new_event(data) }
+          end
+
+        end
+        private_constant :NativeImpl
+
+        module FallbackImpl
+
+          def events_from_json(json, event_factory)
```

Comment:
Is it worth short-cutting to `Event::from_json` for performance if we are working with a `BasicEventFactory`? Doing so would require that `BasicEventFactory` not be a private constant.

```suggestion
          def events_from_json(json, event_factory)
            return Logstash::Event.from_json(json) if event_factory.kind_of?(BasicEventFactory)
```

Reply:
run a few micro benchmarks and the block vs non-block form only differ on small JSON payloads (< 1kB) e.g.
```
json_small (compact) string size: 195
Warming up --------------------------------------
json_small (compact) from_json       
                        33.432k i/100ms
json_small (compact) from_json { ... }
                        30.001k i/100ms
Calculating -------------------------------------
json_small (compact) from_json       
                        360.141k (± 5.4%) i/s -     10.765M in  30.017677s
json_small (compact) from_json { ... }
                        312.449k (± 7.5%) i/s -      9.300M in  30.013306s


json_small (pretty) string size: 287
Warming up --------------------------------------
json_small (pretty) from_json       
                        29.900k i/100ms
json_small (pretty) from_json { ... }
                        27.106k i/100ms
Calculating -------------------------------------
json_small (pretty) from_json       
                        335.784k (± 7.2%) i/s -     10.016M in  30.028366s
json_small (pretty) from_json { ... }
                        298.067k (± 6.2%) i/s -      8.918M in  30.072253s

Comparison:
json_small (pretty) from_json       :   335784.5 i/s
json_small (pretty) from_json { ... }:   298066.8 i/s - same-ish: difference falls within error

```

as the payload gets bigger it's less noticeable:

```
regular string size: 6441
Warming up --------------------------------------
regular from_json       
                         2.611k i/100ms
regular from_json { ... }
                         2.344k i/100ms
Calculating -------------------------------------
regular from_json       
                         26.228k (± 7.2%) i/s -    783.300k in  30.081367s
regular from_json { ... }
                         25.427k (± 5.2%) i/s -    761.800k in  30.070375s

Comparison:
regular from_json       :    26227.5 i/s
regular from_json { ... }:    25426.6 i/s - same-ish: difference falls within error



medium_array string size: 8659
Warming up --------------------------------------
medium_array from_json       
                         1.273k i/100ms
medium_array from_json { ... }
                         1.310k i/100ms
Calculating -------------------------------------
medium_array from_json       
                         12.747k (± 6.3%) i/s -    381.900k in  30.124069s
medium_array from_json { ... }
                         13.317k (± 6.0%) i/s -    398.240k in  30.067693s

Comparison:
medium_array from_json { ... }:    13317.2 i/s
medium_array from_json       :    12747.2 i/s - same-ish: difference falls within error

```
think we're good here for now